### PR TITLE
Prevent conflicting rustup toolchain downloads

### PR DESCRIPTION
**Pull request by kornelski**
Rustup has a shared download folder, and multiple concurrent rustup runs overwrite each others temp files in there, causing sporadic installation failures.

https://github.com/rust-lang/rustup/issues/3690

### CODE REVIEW

File: src/rustup.rs [hunk 1]
```diff
@@ -126,6 +126,10 @@ pub(crate) fn version_range(
     Ok(versions)
 }
 
+/// Prevent concurrent downloads of the same toolchain version
+/// <https://github.com/rust-lang/rustup/issues/3690>
+static RUSTUP_TOOLCHAIN_CHANGES: Mutex<()> = Mutex::new(());
+
 pub(crate) fn install_toolchain(
```
Comment (taiki-e) [inline src/rustup.rs:133]: Hmm, AFAIK there is no case where install_toolchain is called at the same time within a single binary. There are two callers of install_toolchain, both of which are called on the main thread.
https://github.com/taiki-e/cargo-hack/blob/fd670a45280a1a9afd6f7b0fa2d46609e0c6a8a6/src/main.rs#L339
https://github.com/taiki-e/cargo-hack/blob/fd670a45280a1a9afd6f7b0fa2d46609e0c6a8a6/src/rustup.rs#L56

If test-threads are not limited [unlike CI](https://github.com/taiki-e/cargo-hack/blob/fd670a45280a1a9afd6f7b0fa2d46609e0c6a8a6/.github/workflows/ci.yml#L22), there are cases in tests where cargo-hack binaries are called at the same time, but in those cases, locks that are effective only within a single binary may not be useful.

IIUC, the first version of this PR (https://github.com/taiki-e/cargo-hack/commit/d6e5280c29c6edc339de48d37fd47be9d19e0d58) that was trying to limit the concurrency of the tests is more accurate.

**Review by taiki-e — COMMENTED** [review]
Thanks for the PR!

**Comment by kornelski** [comment]
CI fails due to invalid args to `yq` command, which seems unrelated to this PR.

**Comment by taiki-e** [comment]
> CI fails due to invalid args to yq command, which seems unrelated to this PR.

Oh, sorry. I pushed the fix to the main branch now.

**Review by taiki-e — APPROVED** [review]

**Pull request merged**